### PR TITLE
Link to caching documentation page for hooks returning records

### PR DESCRIPTION
**Pull request by sebashwa**
Records returned from a context / controller hook are sometimes `undefined` or a cached version of the record. This is now stated more explicitly in the hook documentations for hooks that return the record from a context.

Things changed:

* Had to move some documentation out of code blocks to add those links to the caching page. 
* Tried to make the documentation for the list / edit / create / show hooks more consistent (added headings / using the same documentation format for Arguments / Return Value). 
* Added "Props"-Headings for some more components to pop up in the navigation

Related to #9451

### CODE REVIEW

File: docs/useCreateContext.md [hunk 1]
```diff
@@ -40,14 +40,10 @@ const PostCreateForm = () => {
...
 
-```jsx
-const {
-    defaultTitle, // the translated title based on the resource, e.g. 'Create New Post'
-    redirect, // the default redirection route. Defaults to 'list'
-    resource, // the resource name, deduced from the location. e.g. 'posts'
-    save, // the update callback, to be passed to the underlying form as submit handler
-    saving, // boolean that becomes true when the dataProvider is called to create the record
-} = useCreateContext();
-```
```
Comment (fzaninotto) [inline docs/useCreateContext.md:53]: I prefer the previous version

File: docs/useCreateController.md [hunk 2]
```diff
@@ -43,24 +43,24 @@ export const BookCreate = () => {
...
 
-```jsx
-const {
-    defaultTitle, // the translated title based on the resource, e.g. 'Create New Post'
-    redirect, // the default redirection route. Defaults to 'list'
-    resource, // the resource name, deduced from the location. e.g. 'posts'
-    save, // the update callback, to be passed to the underlying form as submit handler
-    saving, // boolean that becomes true when the dataProvider is called to create the record
-} = useCreateController();
-```
```
Comment (fzaninotto) [inline docs/useCreateController.md:65]: I prefer this version

File: docs/useEditContext.md [hunk 3]
```diff
@@ -45,23 +45,19 @@ const Aside = () => {
...
-    isLoading, // boolean that is true until the record is available for the first time
-    mutationMode, // mutation mode argument passed as parameter, or 'undoable' if not defined
-    record, // record fetched via dataProvider.getOne() based on the id from the location
-    redirect, // the default redirection route. Defaults to 'list'
-    refetch, // a function that allows you to refetch the record 
-    resource, // the resource name, deduced from the location. e.g. 'posts'
-    save, // the update callback, to be passed to the underlying form as submit handler
-    saving, // boolean that becomes true when the dataProvider is called to update the record
-} = useEditContext();
-```
```
Comment (fzaninotto) [inline docs/useEditContext.md:64]: Same

File: docs/useEditController.md [hunk 4]
```diff
@@ -46,31 +46,31 @@ export const BookEdit = () => {
...
-    resource, // the resource name, deduced from the location. e.g. 'posts'
-    save, // the update callback, to be passed to the underlying form as submit handler
-    saving, // boolean that becomes true when the dataProvider is called to update the record
-} = useEditController();
-```
+## Return Value
+
+`useEditController` returns an object with the following keys:
+
+* `defaultTitle`: Translated title based on the resource, e.g. 'Post #123'
```
Comment (fzaninotto) [inline docs/useEditController.md:66]: keep the object format

File: docs/useList.md [hunk 5]
```diff
@@ -255,39 +256,46 @@ const { data } = useList({
 
 ## Return Value
 
-`useList` returns an object matching the shape of the `ListContext`: 
+`useList` returns an object with keys matching the shape of the `ListContext`: 
 
-```jsx
```
Comment (fzaninotto) [inline docs/useList.md:260]: and the object syntax

File: docs/useListContext.md [hunk 6]
```diff
@@ -56,42 +56,49 @@ export const PostList = () => (
 
 ## Return Value
 
-The `useListContext` hook returns an object with the following keys:
+`useListContext` returns an object with the following keys:
 
-```jsx
```
Comment (fzaninotto) [inline docs/useListContext.md:61]: Please keep the object value

File: docs/useListController.md [hunk 7]
```diff
@@ -141,39 +141,46 @@ You can disable this feature by setting the `storeKey` prop to `false`. When dis
 
 ## Return Value
 
-The return value of `useListController` has the following shape:
```
Comment (fzaninotto) [inline docs/useListController.md:144]: PLease keep the object format

File: docs/useSaveContext.md [hunk 8]
```diff
@@ -41,3 +37,11 @@ const MyComponent = () => {
 }
 ```
 {% endraw %}
+
+## Return Value
+
+`useSaveContext` returns an object with the following keys:
+
+* `save`: Create or update callback which receives form data and calls `dataProvider`
```
Comment (fzaninotto) [inline docs/useSaveContext.md:45]: MAke it a code snippet

File: docs/useShowContext.md [hunk 9]
```diff
@@ -49,6 +43,19 @@ const PostShow = () => (
 )
 ```
 
+## Return Value
+
+`useShowContext` returns an object with the same keys as [`useShowController`](./useShowController.md):
+
+* `defaultTitle`: Translated title based on the resource, e.g. 'Post #123'
```
Comment (fzaninotto) [inline docs/useShowContext.md:50]: same

File: docs/useShowController.md [hunk 10]
```diff
@@ -176,6 +166,18 @@ const PostShow = () => {
 };
 ```
 
+## Return Value
+
+`useShowController` returns an object with the following keys:
+
+* `defaultTitle`: Translated title based on the resource, e.g. 'Post #123'
```
Comment (fzaninotto) [inline docs/useShowController.md:173]: Please use an object in a code snippet

File: docs/useCreateController.md [hunk 11]
```diff
@@ -43,24 +43,24 @@ export const BookCreate = () => {
 
 **Tip**: If you just use the return value of `useCreateController` to put it in an `CreateContext`, use [the `<CreateBase>` component](./CreateBase.md) instead for simpler markup.
 
-`useCreateController` accepts an options argument, with the following fields, all optional:
+## Arguments
```
Comment (fzaninotto) [inline docs/useCreateController.md:46]: We use Parameters in most hooks. Why is Arguments better? And if you change it here, you must change it everywhere.

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Nice! Although I don't agree with all your proposed changes.

File: docs/useCreateController.md (hunk 11 above)
Comment (sebashwa) [reply docs/useCreateController.md:46]: Wasn't sure what's the preference, both work for me. Will change it back soon!

**Comment by sebashwa** [comment]
Hey @fzaninotto,

thanks for your feedback. My intention to use the listing instead of the object syntax in a codeblock were:

* It's more consistent with the documentation for parameters (which are listings in most of the places, although sometimes also tables in case types / default values are documented) and I don't see why the return value should be documented differently from the parameters.
* It's not necessary to scroll to the right to read the descriptions of the object keys (at least I have to, maybe it's a resolution thing)
* I thought it's not possible to put links inside codeblocks and I wanted to link to the "Caching" page for the returned `record` key. This was the main reason why I abandoned the codeblocks, although I think now, that's actually not true.

If you still think I should switch back to codeblocks, I will of course do so but imho I think it's neither more readable nor consistent with the rest of the documentation. What do you think?

**Comment by fzaninotto** [comment]
I hear your arguments, yet I prefer the code snippet for the return values.

**Comment by fzaninotto** [comment]
Any updates?

**Comment by sebashwa** [comment]
> Any updates?

Wasn't able to add those changes earlier.
I reverted to the code block documentation style, although it's now no longer possible to link to the Caching page from within the code block. So it just remains a hint to the Caching page for now.

**Review by fzaninotto — APPROVED** [review]

**Comment by fzaninotto** [comment]
Thanks!

**Pull request merged**